Changes to the first step of the Welcome Wizard [MAILPOET-4815]

rodrigoprimo opened the pull request:
## Description

This PR implements a few changes to the first step of the Welcome Wizard as described in the ticket.

## Code review notes

There is some repetition in `welcome_wizard_controller.jsx` and `woocommerce_controller.jsx` (`finishWizard()` and `updateSettings()`). I did not address this as the repetition was already there before I started this PR and it was not clear to me how to address it in a React way. Please let me know if you have any suggestions and if you think I should change this part.

## QA notes

Go to /wp-admin/admin.php?page=mailpoet-welcome-wizard#/steps/1 to access the first step of the Welcome Wizard and see the changes proposed in this PR.

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-4815]

## After-merge notes

_N/A_


[MAILPOET-4815]: https://mailpoet.atlassian.net/browse/MAILPOET-4815?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

rodrigoprimo committed:
Change copy of the first step of the welcome wizard

[MAILPOET-4815]

rodrigoprimo committed:
Change button in the first step of the welcome wizard

This commit changes the button "Finish later" of the first step of the
welcome wizard to "Skip this step". Both the label and the functionality
of the button changed.

[MAILPOET-4815]

rodrigoprimo committed:
Fix typo

rodrigoprimo committed:
Use DI container to get an instance of Menu to simplify MenuTest

By using the DI container to get an instance of Menu we simplify
MenuTest and avoid the need to change MenuTest::getMenu() every time the
signature of the constructor of the Menu class changes, which will
happen in a subsequent commit.

[MAILPOET-4815]

rodrigoprimo committed:
Force users to complete the Welcome Wizard

This commit changes the behavior of the Welcome Wizard. Up until now,
the wizard would load once and users could opt to complete or skip it.
Now it is mandatory for the users to complete the wizard. Before we
saved the setting "version" the first time the first step of the wizard
was loaded and now we saved it when the wizard is completed.

[MAILPOET-4815]

rodrigoprimo committed:
Display submenu entries only after the Welcome Wizard is completed

[MAILPOET-4815]

rodrigoprimo committed:
Change ReinstallFromScratch test now that the Welcome Wizard is mandatory

This commit updates the ReinstallFromScratch test now that the logic for
displaying the Welcome Wizard changed in
afba41ad27b8b72e892892fe386b48b5a6f59b9b. Before just loading the wizard
once was enough to dismiss it, now it is necessary to complete it. In
this commit, an utility function was created and it is used to dismiss
the wizard twice in the modified test.

[MAILPOET-4815]

rodrigoprimo committed:
Refactor sender_step.jsx to TypeScript

[MAILPOET-4815]

rodrigoprimo committed:
Refactor welcome_wizard_controller.jsx to TypeScript

[MAILPOET-4815]

rodrigoprimo committed:
Refactor wizard.jsx to TypeScript

[MAILPOET-4815]

rodrigoprimo committed:
Refactor webpack_admin_index.jsx to TypeScript

[MAILPOET-4815]

rodrigoprimo commented:
@samnajian, I refactored three of the four JSX files that were changed in this PR to TSX. I didn't refactor `woocommerce_controller.jsx` as I plan to touch this file on #4605. Could you please review this PR again?

Please note that this is one of the first times that I have refactored JSX files to TSX, so I'm not sure about all the changes that I made and it is likely I missed something.

samnajian commented on mailpoet/assets/js/src/wizard/steps/sender_step.tsx:
```diff
@@ -52,47 +72,32 @@ function WelcomeWizardSenderStep(props) {
               type="text"
               placeholder="john@doe.com"
-              value={props.sender ? props.sender.address : ''}
+              value={sender ? sender.address : ''}
```
You can replace this with [optional chaining](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining) as well

rodrigoprimo replied:
That is good to know. I don't think it is necessary to create a commit to change it in this case, but I will try to remember next time. Let me know if you disagree and you think I should change it in this PR.

samnajian replied:
It's fine to leave it like this.

samnajian commented on mailpoet/assets/js/src/wizard/welcome_wizard_controller.tsx:
```diff
@@ -71,31 +81,29 @@ function WelcomeWizardStepsController(props) {
       e.preventDefault();
       setLoading(true);
-      updateSettings(
+      void updateSettings(
```
Not sure what value does this `void` operator add in these case, ts seems like to be happy without it as well, and I guess removing it doesn't change the logic either.

rodrigoprimo replied:
Without the `void` ESLint returns the following error:

```
mailpoet/mailpoet/assets/js/src/wizard/welcome_wizard_controller.tsx
  63:5  error  Promises must be awaited, end with a call to .catch, end with a call to .then with a rejection handler or be explicitly marked as ignored with the `void` operator  @typescript-eslint/no-floating-promises
```

My assumption is that since we are already handling errors inside `updateSettings()`, it is safe to use `void` here instead of using `catch()` or the second parameter of `then()`, but I'm not 100% sure. What do you think?

samnajian replied:
I'd prefer prefixing with `await` to better expose the exceptions if they happen specially since as a user we don't know if and how the exceptions are handled inside `updateSettings`, besides this function `submitTracking` is stalling the application by setting `setLoading(true);` so by nature it's sync and it's fine to await it.

rodrigoprimo replied:
I don't know if I'm missing something as I'm not super familiar the all the concepts here, but I don't think I can simply use `await` as it should be used only inside an async function or at the top level of a module. Neither is the case for the places where I used `void`. Happy to chat synchronously about this tomorrow if it is easier so that we can find a solution.

rodrigoprimo replied:
Thanks for the information you shared on Slack, Sam. This PR is ready for a final review.

samnajian reviewed (commented):
Thanks for the great work and js->ts conversions as well, I've just added some minor comments.

rodrigoprimo committed:
Replace void with async/await based on input from PR reviewer

See https://github.com/mailpoet/mailpoet/pull/4603#discussion_r1048950706

[MAILPOET-4815]

samnajian reviewed (dismissed):
LGTM, thanks for the updates.

Aschepikov reviewed (dismissed):
GJ!

Aschepikov commented:
@rodrigoprimo Hi, There are some conflicts can you pls resolve them

samnajian committed:
Fix issue in wizard.tsx with using Route.render

Route.render expects other types than the one coming out of
withBoundary, replacing it with Route.component fixes the
typing issue

[MAILPOET-4815]

rodrigoprimo committed:
Fix ESLint errors after rebasing PR

Use async/await for promisses. Similar to what was discussed here: https://github.com/mailpoet/mailpoet/pull/4603#discussion_r1048950706

This is the ESLint error that started happening after the rebase:

```
mailpoet/assets/js/src/wizard/woocommerce_controller.tsx
  71:11  error  Promises must be awaited, end with a call to .catch, end with a call to .then with a rejection handler or be explicitly marked as ignored with the `void` operator  @typescript-eslint/no-floating-promises
```

[MAILPOET-4815]

rodrigoprimo commented:
@samnajian, could you please take a final look at this PR after the rebasing and the subsequent fixes? Thanks.

samnajian reviewed (approved):
LGTM! thanks for the updates

Aschepikov reviewed (approved):
Verified, gj!
